### PR TITLE
[ExprConstant] Handle shift overflow the same way as other kinds of overflow

### PR DESCRIPTION
**Pull request by efriedma-quic**
We have a mechanism to allow folding expressions that aren't ICEs as an extension; use it more consistently.

This ends up causing bad effects on diagnostics in a few cases, but that's not specific to shifts; it's a general issue with the way those uses handle overflow diagnostics.

### CODE REVIEW

File: clang/test/AST/Interp/shifts.cpp [hunk 1]
```diff
@@ -200,14 +200,7 @@ namespace LongInt {
...
-    X2 = (-1<<29), // cxx17-error {{expression is not an integral constant expression}} \
-                   // cxx17-note {{left shift of negative value -1}} \
-                   // ref-cxx17-error {{expression is not an integral constant expression}} \
-                   // ref-cxx17-note {{left shift of negative value -1}}
-
-    X3 = (1<<32) // all-error {{expression is not an integral constant expression}} \
-                 // all-note {{shift count 32 >= width of type 'int'}}
+    X = (1<<-29),
+    X2 = (-1<<29),
+    X3 = (1<<32),
```
Comment (zygoloid) [inline clang/test/AST/Interp/shifts.cpp:205]: It's concerning that we don't produce a warning by default for constant-folding this, especially given that GCC does not constant-fold bad shifts. I'm not sure it's OK that we start silently accepting this invalid code by default. I wonder if we could reasonably refuse to constant-fold if the evaluation had UB?

File: clang/test/CXX/basic/basic.types/p10.cpp [hunk 2]
```diff
@@ -142,7 +142,7 @@ constexpr int arb(int n) { // expected-note {{declared here}}
                expected-note {{function parameter 'n' with unknown value cannot be used in a constant expression}}
 }
 constexpr long Overflow[(1 << 30) << 2]{}; // expected-warning {{requires 34 bits to represent}} \
-                                              expected-warning {{variable length array folded to constant array as an extension}} \
+                                              expected-error {{variable length array declaration not allowed at file scope}} \
```
Comment (zygoloid) [inline clang/test/CXX/basic/basic.types/p10.cpp:145]: Hm. Why do we stop folding here but start folding in enumerator values?

**Comment by zygoloid** [comment]
As noted in one of the inline comments, it doesn't seem good to be losing a diagnostic for ill-formed code. But this all makes sense to me as a step to make our behavior more consistent. I wonder if it'd make sense to make our folding behavior stricter by default as a prerequisite.

File: clang/test/CXX/basic/basic.types/p10.cpp (hunk 2 above)
Comment (efriedma-quic) [reply clang/test/CXX/basic/basic.types/p10.cpp:145]: It depends on what the caller asks for.  TryToFixInvalidVariablyModifiedType calls EvaluateAsInt directly in a mode that doesn't allow undefined behavior, so it's just treated as an unevaluatable expression.  A few other places use CheckConvertedConstantExpression, which is also strict.  Enumerators go through VerifyIntegerConstantExpression, which is not strict.

File: clang/test/AST/Interp/shifts.cpp (hunk 1 above)
Comment (efriedma-quic) [reply clang/test/AST/Interp/shifts.cpp:205]: We could try messing with the behavior of VerifyIntegerConstantExpression, I guess.

File: clang/lib/AST/ExprConstant.cpp [hunk 3]
```diff
@@ -2849,19 +2849,23 @@ static bool handleIntIntBinOp(EvalInfo &Info, const BinaryOperator *E,
...
       // operand, and must not overflow the corresponding unsigned type.
       // C++2a [expr.shift]p2: E1 << E2 is the unique value congruent to
       // E1 x 2^E2 module 2^N.
-      if (LHS.isNegative())
+      if (LHS.isNegative()) {
         Info.CCEDiag(E, diag::note_constexpr_lshift_of_negative) << LHS;
-      else if (LHS.countl_zero() < SA)
+        if (!Info.noteUndefinedBehavior())
+          return false;
+      } else if (LHS.countl_zero() < SA) {
```
[the remark this answers is not in the document]
Comment (efriedma-quic) [reply clang/lib/AST/ExprConstant.cpp:2863]: This code is under a `LHS.isSigned() && !Info.getLangOpts().CPlusPlus20` guard.

**Comment by efriedma-quic** [comment]
Pushed new version with an update to make Sema::VerifyIntegerConstantExpression slightly more strict.  That handles the cases where the errors were getting dropped.  The tradeoff is that we're more strict about arithmetic overflow in certain integer constant expressions (clang/test/SemaCXX/enum.cpp).

**Review by shafik — COMMENTED** [review]
I think I am bit surprised more tests are not affected by this change considering you modified the handling in several places. It seems we should be adding more test coverage with this change.

File: clang/test/CXX/basic/basic.types/p10.cpp (hunk 2 above)
Comment (AaronBallman) [reply clang/test/CXX/basic/basic.types/p10.cpp:145]: Perhaps we should file an issue to come back and improve consistency here in the future? The behavior is likely mysterious from a user perspective.

In terms of losing the extension, I think it's unlikely to cause significant disruption for anyone so I don't think it needs to be addressed immediately.

**Review by AaronBallman — COMMENTED** [review]
Thank you for working on this! In general, this is heading in the right direction. Though I do agree with @shafik that it would be nice to add additional test coverage for the changes which didn't modify existing test behavior. Also, we should have a release note for the changes and probably a mention about it being a potentially breaking change because we're rejecting code we used to explicitly accept as an extension.

File: clang/test/CXX/basic/basic.types/p10.cpp (hunk 2 above)
Comment (efriedma-quic) [reply clang/test/CXX/basic/basic.types/p10.cpp:145]: The diagnostic shouldn't be hard to fix; it's just a matter of using the form of Evaluate() that returns notes.

For the actual behavior here... we could pass the flag to allow undefined behavior when we do this folding, I guess, but I'd rather not.  Or if we do, it should be a DefaultError diagnostic.

**Comment by efriedma-quic** [comment]
Filed #99680 and #99684 for issues caused by custom codepaths for evaluating ICEs.  Hopefully we can improve that at some point.

**Comment by efriedma-quic** [comment]
Added a few more tests; let me know if there are other tests I should be considering.

**Review by shafik — COMMENTED** [review]
Thank you for the additional tests, so I am not sure if the coverage is sufficient I see changed under the following diagnostic:

note_constexpr_negative_shift
note_constexpr_large_shift
note_constexpr_lshift_of_negative
note_constexpr_lshift_discards

but I don't think I see tests that cover the negative shift changes, am I just missing it?

**Comment by efriedma-quic** [comment]
There's some existing coverage of the affected diagnostics, which shows we continue to produce those diagnostics... but I added a few more tests for in-class static member variables.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by shafik** [comment]
> There's some existing coverage of the affected diagnostics, which shows we continue to produce those diagnostics... but I added a few more tests for in-class static member variables.

Thank you!

**Pull request merged**